internal/c/parts/video/font/font.cpp

```diff
@@ -1,339 +1,1417 @@
+            7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7, // a0..bf
+            8,8,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2, // c0..df
+            0xa,0x3,0x3,0x3,0x3,0x3,0x3,0x3,0x3,0x3,0x3,0x3,0x3,0x4,0x3,0x3, // e0..ef
+            0xb,0x6,0x6,0x6,0x5,0x8,0x8,0x8,0x8,0x8,0x8,0x8,0x8,0x8,0x8,0x8, // f0..ff
+            0x0,0x1,0x2,0x3,0x5,0x8,0x7,0x1,0x1,0x1,0x4,0x6,0x1,0x1,0x1,0x1, // s0..s0
+            1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,1,1,1,1,1,0,1,0,1,1,1,1,1,1, // s1..s2
+            1,2,1,1,1,1,1,2,1,2,1,1,1,1,1,1,1,1,1,1,1,1,1,2,1,1,1,1,1,1,1,1, // s3..s4
+            1,2,1,1,1,1,1,1,1,2,1,1,1,1,1,1,1,1,1,1,1,1,1,3,1,3,1,1,1,1,1,1, // s5..s6
+            1,3,1,1,1,1,1,3,1,3,1,1,1,1,1,1,1,3,1,1,1,1,1,1,1,1,1,1,1,1,1,1, // s7..s8
+        };
+        // clang-format on
+
+        uint32_t type = utf8d[byte];
+        *codep = (*state != UTF8DecoderState::ACCEPT) ? (byte & 0x3fu) | (*codep << 6) : (0xff >> type) & (byte);
+        *state = utf8d[256 + *state * 16 + type];
+
+        return *state;
+    }
 
-        static int32_t glyph_index;
-        glyph_index = FT_Get_Char_Index(fonts[i].handle, codepoint_value);
-        if (!glyph_index) {
-            // failed!
+    /// @brief Resizes the UTF32 conversion buffer
+    /// @param count New codepoints required
+    /// @return True if the buffer was allocated correctly
+    bool ResizeCodepointBuffer(FT_ULong count) {
+        if (count <= codepoints || !count) {
+            codepoints = count;
+            return codepoint != nullptr;
         }
-        if (FT_Load_Glyph(fonts[i].handle, glyph_index, FT_LOAD_DEFAULT)) {
-            // failed!
+
+        auto *tempBuffer = (FT_ULong *)realloc(codepoint, count * sizeof(FT_ULong));
+        if (!tempBuffer)
+            return false;
+        codepoint = tempBuffer;
+
+        FONT_DEBUG_PRINT("UTF32 conversion buffer resized from %i to %i", codepoints, count);
+
+        codepoints = count;
+
+        return true;
+    }
+
+    /// @brief Converts an ASCII array to UTF-32
+    /// @param str The ASCII array
+    /// @param byte_len The size of the array in bytes
+    /// @return The number of codepoints that were converted
+    FT_ULong ConvertASCII(const FT_Bytes str, FT_ULong byte_len) {
+        // Resize the codepoint buffer
+        if (ResizeCodepointBuffer(byte_len)) {
+            // Convert the ASCII string
+            for (FT_ULong i = 0; i < byte_len; i++)
+                codepoint[i] = codepage437_to_unicode16[str[i]];
+
+            return byte_len;
         }
 
-        if (monochrome) {
+        return 0;
+    }
+
+    /// @brief Converts an UTF-8 array to UTF-32. This does not check for BOM
+    /// @param str The UTF-8 array
+    /// @param byte_len The size of the array in bytes
+    /// @return The number of codepoints that were converted
+    FT_ULong ConvertUTF8(const FT_Bytes str, FT_ULong byte_len) {
```

Comment:
Well, we need not use these FT types at all. I was just using those to avoid type casting and compiler warnings. But it looks like after using these, I ended up type casting more than I thought. 😁 I'll be changing all of these to `stdint` types. That'll also make it easier if we ever decide to swap out the True Type library in the future.

Reply:
Replaced most `FT_Byte`, `FT_Bytes` and `FT_ULong` with `stdint` types.